Ubuntu 22.04 & Python 3.10

keithbauer opened the pull request:
## Description
<!--- Describe your changes -->
Brings the stack up to date with current packages. Update Ubuntu from 18.04 to 22.04 and Python from 3.6 to 3.10. Combine and upgrade the requirements from core into the respective requirements in the root.

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Update to Python 3.10 since 3.6 is now EOL. Update requirements so they don't fall too far behind and address depreciations. Most of the requirements were easily upgradable.

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Some failing tests need to be fixed.

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

keithbauer committed:
Combine requirements from core and upgrade

keithbauer committed:
Remove `tox.ini` from core

No longer used after the core merge. Everything in here is now in the main `tox.ini`.

keithbauer committed:
Update `postgres://` to `postgresql://`

keithbauer committed:
Remove semicolons

keithbauer committed:
Update uszipcode db

keithbauer committed:
Upgrade to Ubuntu 22.04 LTS

keithbauer committed:
Update node to 14.x

10.x is not supported on Ubuntu 22.04

keithbauer committed:
Add pkg-config to apt installs

xmlsec requires this

keithbauer committed:
Combined requirements

keithbauer committed:
Update Python to 3.10

Upgrade to 3.10 if possible, but can downgrade if needed

keithbauer committed:
Upgrade pip and setuptools in addition to wheel

keithbauer committed:
Remove uWSGI

Docker setup now uses gunicorn and supervisor

keithbauer committed:
Update `select()`

keithbauer committed:
`uszipcode` added fuzzy search that will raise `ValueError` for an invalid city/state combo

Fuzzy search cannot be disabled

keithbauer committed:
Remove `None` tzinfo from test

As of Flask 2.0 the datetime object is timezone-aware. Removing this `replace()` fixes a `TypeError: can't subtract offset-naive and offset-aware datetimes`

keithbauer committed:
Add a `ValueError` catch for bad date format

Resolves ValueError: Invalid date value or format "01 Aug 2019” raised by email.utils.parsedate_to_datetime

keithbauer committed:
Move collection assignment

Resolves a TypeError raised by tests

keithbauer committed:
Compile is now ran in the constructor

keithbauer committed:
Update `Binary` format to `LargeBinary`

`Binary` was deprecated and the correct format now is `LargeBinary`

keithbauer committed:
Update `._distinct` to `._distinct_on`

keithbauer committed:
Update jwt `algorithm` to `algorithms`

keithbauer committed:
Remove jwt decode

Is now a `str` object

keithbauer committed:
Encode the jwt encode

Expects a bytes-like object, not str

keithbauer committed:
Update expected SAML XML

I think this is for the `certifi` update

keithbauer committed:
Test with 3.10

Could include more versions, maybe 3.7+

keithbauer committed:
Put python version in quotes

keithbauer committed:
Update python 3.6 to 3.10

keithbauer commented on Dockerfile:
```diff
@@ -75,15 +75,16 @@ RUN apt-get update \
     language-pack-en \
     git \
-    python3.6 \
+    python3.10 \
```
If we don't feel comfortable moving to 3.10 yet then this could be downgraded.

keithbauer commented on api/admin/geographic_validator.py:
```diff
@@ -74,27 +74,31 @@ def validate_geographic_areas(self, values, db):
+                        else:
+                            # Flag this as needing to be checked with the registry
+                            flagged = True
+                    except ValueError:
```
`uszipcode` added fuzzy searching that can't be toggled off. An invalid city/state search will attempt to find something while raising a `ValueError`.

keithbauer commented on requirements-base.txt:
```diff
@@ -1,61 +1,118 @@
+# watchtower is for Cloudwatch logging integration
+watchtower==0.8.0
 wcag-contrast-ratio==0.9
+webpub-manifest-parser==0.0.4
```
This package needs an [update for the path](https://github.com/NYPL-Simplified/python-webpub-manifest-parser/blob/master/src/webpub_manifest_parser/core/registry.py#L25) to `MutableMapping`. Instead of `collections.MutableMapping` it is now located at `collections.abc.MutableMapping`.

leonardr replied:
You can fix this yourself pretty easily with a PR. However, the only developer with PyPI upload permissions is a former Lyrasis contractor who lives in Russia, so it might be tricky to contact him. 

Since we're not planning on using this for anything else, I think it'd be OK to copy this into the circulation repo and make the change there, at least as a temporary measure.

keithbauer commented on requirements.txt:
```diff
@@ -1,2 +1,4 @@
+# All the requirements should be listed in requirements-base.txt except
+# for psycopg2, which is listed here.
 -r requirements-base.txt
+psycopg2==2.8.6
```
I think I've put this in the right place after combining requirements. Same for `psycopg2-binary`.

keithbauer commented on tests/saml/configuration/test_model.py:
```diff
@@ -499,22 +499,24 @@ def test_get_settings_returns_correct_result(self):
                 "wantAssertionsEncrypted": False,
                 "nameIdEncrypted": False,
-                "signatureAlgorithm": "http://www.w3.org/2000/09/xmldsig#rsa-sha1",
+                "signatureAlgorithm": "http://www.w3.org/2001/04/xmldsig-more#rsa-sha256",
```
I think upgrading the `certifi` package updated these fields. I'm not very familiar with the SAML configuration, but hopefully this is ok.

leonardr replied:
Yeah, it should be, SHA1 has been deprecated in favor of SHA256 in general.

keithbauer commented on tests/test_opds.py:
```diff
@@ -1015,7 +1015,7 @@ def test_active_loan_feed(self):
         # strips timezone information from it,
         # so to verify we have the right value we must do the same.
-        last_sync_naive = patron.last_loan_activity_sync.replace(tzinfo=None)
+        last_sync_naive = patron.last_loan_activity_sync
```
As of Flask 2.0 the datetime object is timezone-aware

keithbauer commented:
At the moment the tests fail due to the import error for `MutableMapping` from `webpub_manifest_parser`, which won't expose the additional errors from upgrades. I edited this locally so I could continue with the upgrades and fixes.

keithbauer commented on requirements-base.txt:
```diff
@@ -1,61 +1,118 @@
+cryptography==37.0.2
 # defusedxml is required for SAML authentication
-defusedxml==0.6.0
-dm.xmlsec.binding==2.1
```
I don't think `dm.xmlsec.binding` is required now with `python3-saml`. I think it was required `python-saml`. If that's not the case though then this will need to wait until support for versions of Python > 3.6 is pushed to PyPI.

leonardr replied:
That's really good news, because these two packages were the reason we couldn't upgrade Python earlier.

leonardr commented on api/adobe_vendor_id.py:
```diff
@@ -801,7 +801,7 @@ def _encode(self, iss=None, sub=None, iat=None, exp=None):
             payload['exp'] = self.numericdate(exp) # Expiration Time
         return base64.encodebytes(
-            jwt.encode(payload, self.secret, algorithm=self.ALGORITHM)
+            jwt.encode(payload, self.secret, algorithm=self.ALGORITHM).encode()
```
Did jwt change such that encode now returns a Unicode string?

keithbauer replied:
Yes, with [v2.0.0](https://pyjwt.readthedocs.io/en/stable/changelog.html#jwt-encode-return-type) a string is returned instead of a byte string.

leonardr commented on core/metadata_layer.py:
```diff
@@ -805,6 +805,10 @@ def mirror_link(self, model_object, data_source, link, link_obj, policy):
+        # If the collection is accessed later than here
+        # a TypeError: can't escape str to binary is raised
+        collection = pools[0].collection if pools else None
+
```
This is weird.

keithbauer replied:
This one was very strange and I don't know why the TypeError was raising or why this fixes it...

leonardr commented on core/model/resource.py:
```diff
@@ -565,7 +565,7 @@ class Representation(Base, MediaTypes):
 
     # The content of the representation itself.
-    content = Column(Binary)
+    content = Column(LargeBinary)
```
Do these turn into the same underlying postgresql data type as before?

keithbauer replied:
Yes this is the same data type as before. https://docs.sqlalchemy.org/en/14/changelog/changelog_14.html#change-214869a306f72f33237772d33fc332ec

leonardr commented on docker/docker-compose.yml:
```diff
@@ -36,8 +36,8 @@ services:
-      SIMPLIFIED_PRODUCTION_DATABASE: postgres://simplified:password@db:5432/simplified_circulation_dev
-      SIMPLIFIED_TEST_DATABASE: postgres://simplified_test:password@db:5432/simplified_circulation_test
+      SIMPLIFIED_PRODUCTION_DATABASE: postgresql://simplified:password@db:5432/simplified_circulation_dev
+      SIMPLIFIED_TEST_DATABASE: postgresql://simplified_test:password@db:5432/simplified_circulation_test
```
It occurs to me we're going to have to mirror this change in our build processes, since that's where the environment variables come from.

keithbauer committed:
Bring in webpub_manifest_parser

keithbauer committed:
Use Python 3.9

keithbauer committed:
Remove webpub-manifest-parser from requirements

Is now brought in under `core/util`

keithbauer committed:
Add requirements for webpub-manifest-parser

keithbauer committed:
Update `._limit` to `._limit_clause`

keithbauer committed:
Update `logging.warn()` to `logging.warning()`

logging.warn has been deprecated

keithbauer committed:
Enable warnings and verbosity

keithbauer committed:
Fix verbosity flag

keithbauer committed:
Update `.__to__string()` to `render_as_string`

URL.__to_string__  is deprecated, can use render_as_string with hide_password=True

keithbauer committed:
Update `cache_timeout` to `max_age`

cache_timeout will be removed in Flask 2.2

keithbauer commented:
This error occurs a few times and is one I don't really understand.

- `AssertionError: Collection was loaded during event handling.`: These errors seem to be raised when assigning works to licensepools. Putting the assignments in a try/except block and re-trying the assignment in the except block seems to get them to pass, but this is not a good solution.

I also enabled warnings for the most recent test outputs and noticed some additional SQLAlchemy warnings, I'm not sure if they're really issues:

- `SAWarning: Coercing Subquery object into a select() for use in IN(); please pass a select() construct explicitly`. This occurs in the Work when generating search documents [here](https://github.com/NYPL-Simplified/circulation/blob/develop/core/model/work.py#L1659) and [here](https://github.com/NYPL-Simplified/circulation/blob/develop/core/model/work.py#L1689).
- `SAWarning: SELECT statement has a cartesian product between FROM element(s) ... Apply join condition(s) between each element to resolve.` This also occurs when [generating the search document](https://github.com/NYPL-Simplified/circulation/blob/develop/core/model/work.py#L1766) and [querying for the ElasticSearch integration](https://github.com/NYPL-Simplified/circulation/blob/develop/api/app.py#L54).
- `SA Warning: relationship X will copy column Q to column P, which conflicts with relationship(s): ‘Y’` [This test](https://github.com/NYPL-Simplified/circulation/blob/develop/core/tests/models/test_resource.py#L566) displays several relationship warnings and the warning message has a link to [this resource](https://github.com/NYPL-Simplified/circulation/blob/ERD-36/core/tests/models/test_resource.py#L566) and mentions configuring an overlaps parameter to some relationships.

I also noticed that there were some `PytestCollectionWarning cannot collect test class '<TestClass>' because it has a __init__ constructor (from: <test_file.py>)`. There are 10 of these warnings and it does look like these tests were not collected or ran. I'll dig into this further since every test should run.

keithbauer commented:
PyJWT made some changes that affect the tests for `ExpiredSignatureError` and `InvalidIssuedAtError`. This [pull request](https://github.com/jpadilla/pyjwt/pull/252/commits/8f3a2a8a4098693357b69d63a1dbec514ed7c701) no longer rejects tokens with a future `iat` value. [This test](https://github.com/NYPL-Simplified/circulation/blob/develop/tests/test_adobe_vendor_id.py#L803) and [this test](https://github.com/NYPL-Simplified/circulation/blob/develop/tests/test_adobe_vendor_id.py#L812) now raise a `DecodeError: Unknown library: Patron identifier` instead of their respective errors, should these tests raise a `DecodeError` instead?

Looking at the [documentation](https://pyjwt.readthedocs.io/en/latest/api.html#jwt.decode), `iat` and `exp` are only verified if present in a `require` key passed in the `options` dict. It doesn't look like they'll be verified if `verify_signature=False`, which is set [here](https://github.com/NYPL-Simplified/circulation/blob/develop/api/adobe_vendor_id.py#L866), unless `verify_iat` or `verify_exp` is explicitly set to True and passed in as part of the `options` otherwise they are [set as False](https://github.com/jpadilla/pyjwt/blob/master/jwt/api_jwt.py#L88).

keithbauer commented:
After looking into the `PytestCollectionWarning` warnings, I think they're ok. All the tests are being ran. I think the collection warning is due to some test classes being used as a sort of mock class that don't have test methods, e.g. `TestLicense` or `TestLicenseInfo` used in `TestODLExpiredItemsReaperSingleLicense`.

leonardr commented:
I recently realized that the entire adobe_vendor_id.py file is obsolete ever since we moved Open eBooks off of Adobe DRM in 2021. Rather than spend a lot of time trying to figure out the jwt test failures, we could remove that whole file and its tests, and potentially the jwt dependency. However, similar code is present in the library registry so we might run into the problem again and have to solve it for real.

I guess I could live with an implementation that allowed for a future `iat`, since I don't think that case would realistically happen, even if this code was still in use. But a token that's more than an hour old shouldn't be accepted.

Talk to Risa about removing the adobe_vendor_id.py file altogether, rather than fixing those bugs. I don't think we'll ever need that code again, and if we do we can model it after the code in the library registry.

leonardr commented:
If you end up removing that file, do it in a separate PR that lands before this one does.

leonardr commented:
I don't have a clue about the AssertionError while adding something to a collection. The method where the assertion is failing hasn't changed recently.

It looks like it's got something to do with this relationship on Work:

```
license_pools = relationship("LicensePool", backref="work", lazy='joined')
```

Setting LicensePool.work adds the LicensePool to this list, and that ends up in the bit of code that's failing.

As a way to get more information I'd try first changing the `lazy` option, which is a performance optimization, and see if that solves the problem. Then I'd put a breakpoint inside that append method and look at all the state, especially `self.key`, the thing that's not supposed to be in `dict_` but apparently is.

keithbauer committed:
Add timeout parameter for SAML urlopen tests

keithbauer commented:
Would  `adobe_vendor_id.py` still be needed for SimplyE if it is serving Adobe DRM content?

leonardr commented:
I don't think so -- adobe_vendor_id.py is an implementation of Adobe's Authentication Web Service, and we've told Adobe to use the library registry implementation at libraryregistry.librarysimplified.org instead.

To test this out, you could check the logs to if there are any hits coming in to /AdobeAuth/* on circulation.librarysimplified.org. If there are any, we probably need to have a talk with Adobe. If not, we should be good to remove those endpoints and all the code behind them.

keithbauer commented:
I looked through the logs for circulation manager and didn't see any instances of /AdobeAuth/*, but did see them in the logs for registry so I also think this is safe to remove now. I'll work on that in a separate PR.

keithbauer committed:
Bump Python to 3.10

Ubuntu 22.04 does not have 3.9

keithbauer committed:
Bump Pillow to 9.1.1

keithbauer committed:
Bump Python to 3.10

keithbauer committed:
Append the LicensePool to the Work

Instead of assigning the Work directly to the LicensePool's work. This gets around a SQLAlchemy AssertionError: Collection was loaded during event handling.

keithbauer committed:
Merge branch 'develop' into ERD-36

keithbauer committed:
Fix missing requirements for apispec

Missed this when fixing the merge conflict

keithbauer committed:
Mock url_for in MockManager

Using url_for with redirects was problematic since url_for was getting mocked through MockControllerMethod which was returning a Response, but redirect() requires a string and not a Response.

keithbauer committed:
Fix search by county

keithbauer committed:
Fix fast_query_count

Get the distinct columns from statement._distinct_on and the table from the query statement. Construct a new select with the columns and table  to get the count.

keithbauer committed:
Fix query limit check and return value

keithbauer committed:
Parametrize process_cover_image test

keithbauer committed:
Update blue_with_title_author.png

The previous image has enough subtle differences to cause the root mean square to deviate. This new image has been generated from the latest version of Pillow.

keithbauer committed:
Explicitly delete work

Normally this is done through collection.delete(), but due to writing database commits at a determined frequency this work's license pool's deletion does not trigger the work's self deletion.

keithbauer committed:
Disable test warnings

keithbauer committed:
Use Python 3.10 for docs

keithbauer committed:
Use ubuntu-22.04 for GitHub Actions

keithbauer committed:
Remove tests verbose flag

keithbauer committed:
Install libjpeg-dev

See if this resolves the process cover image test

keithbauer committed:
Save cover images as artifacts for analysis

keithbauer committed:
Use upload-artifact not checkout...

keithbauer committed:
Always run the cover image artifact upload

keithbauer committed:
Bump actions checkout version

keithbauer committed:
Bump actions setup-python version

keithbauer committed:
Try installing libfreetype-dev to resolve font issues for test_process_cover_image

Something is causing fonts to render slightly differently in GitHub Actions vs our docker images, this is an attempt to resolve that.

keithbauer committed:
Use libfreetype-dev

keithbauer committed:
Try specifying libraqm-dev

keithbauer committed:
Stop tests after first failure

Revert this when there's a solution for the cover image text

keithbauer committed:
Try a different character for char width

Grasping at straws here...

keithbauer committed:
Revert "Try a different character for char width"

This reverts commit 07228beef404625c099950aac75cf975a8d642dd.

keithbauer committed:
Remove libraqm-dev libjpeg-dev from GH Actions

keithbauer committed:
Try math.floor() for rectangle

It looks to be a few pixels too wide in CI

keithbauer committed:
Revert math.floor()

keithbauer committed:
Revert fail fast

keithbauer committed:
Removing saving cover images in CI

keithbauer committed:
Remove artifact upload for cover images

Finished troubleshooting test_process_cover_image

keithbauer committed:
Adjust root mean square tolerance

keithbauer committed:
Remove saving cover images

Was used to debug issue in CI

leonardr commented on requirements-base.txt:
```diff
@@ -1,64 +1,125 @@
 unicodecsv==0.14.1
 uritemplate==3.0.1
-uszipcode==0.2.4
+# urllib3 is pinned to a pre-1.25 version to work around https://jira.nypl.org/browse/SIMPLY-3477
```
This comment can now be removed.

leonardr reviewed (approved):
Looks good to me.

keithbauer committed:
Remove outdated comment
